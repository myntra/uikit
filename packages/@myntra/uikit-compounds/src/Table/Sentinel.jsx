import React, { PureComponent, Fragment } from 'react'
import PropTypes from 'prop-types'
import { Measure } from '@myntra/uikit-elements'
import { classnames } from '@myntra/uikit-utils'

// eslint-disable-next-line
const FragmentWithFallback = Fragment || (({ children }) => <div style={{ display: 'contents' }}>{children}</div>)

/**
 * A component that renders a sentinel node with same size as the reference node.
 *
 * @since 0.5.0
 * @status EXPERIMENTAL
 * @example
 * <Sentinel />
 */
export default class Sentinel extends PureComponent {
  static propTypes = {
    /** @private */
    className: PropTypes.string,
    children: PropTypes.node.isRequired,
    transform: PropTypes.func,
    transformSentinel: PropTypes.func
  }

  static defaultProps = {
    transform({ height }) {
      return { marginTop: `-${height}` }
    },
    transformSentinel({ height }) {
      return { height }
    }
  }

  state = {
    size: {
      height: '0px',
      width: '0px'
    }
  }

  handleResize = ({ bounds: { width, height } }) => {
    width = width + 'px'
    height = height + 'px'

    if (width !== this.state.size.width || height !== this.state.size.height) {
      this.setState({ size: { width, height } })
    }
  }

  render() {
    const node = React.Children.only(this.props.children)

    return (
      <FragmentWithFallback>
        <div style={this.props.transformSentinel(this.state.size)} key="sentinel" />
        <Measure bounds onMeasure={this.handleResize} key="measure">
          {React.cloneElement(node, {
            style: { ...node.props.style, ...this.props.transform(this.state.size) },
            className: classnames(this.props.className, node.props.className)
          })}
        </Measure>
      </FragmentWithFallback>
    )
  }
}
