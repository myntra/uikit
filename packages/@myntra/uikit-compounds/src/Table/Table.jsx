import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { memoize } from '@myntra/uikit-utils'
import { prepareHead, prepareBody } from './TableHelpers'

import TableColumn from './TableColumn'
import TableSimple from './TableSimple'
import { Sortable } from '@myntra/uikit-elements'

/**
A simple table.

 @since 0.3.0
 @status EXPERIMENTAL
 @example
<Table data={[
    { id: 1, firstName: "Jane", lastName: "Doe" },
    { id: 2, firstName: "John", lastName: "Doe" },
    { id: 3, firstName: "Adam", lastName: "Seed" },
  ]} sort={['firstName']}>
  <Table.Column key="id" label="ID">
    {({ data }) => <div>{data.id}</div>}
  </Table.Column>
  <Table.Column key="firstName" label="First Name">
    {({ data }) => <div>{data.firstName}</div>}
  </Table.Column>

  <Table.Column key="lastName" label="Last Name">
    {({ data }) => <div>{data.lastName}</div>}
  </Table.Column>
</Table>
 */
class Table extends PureComponent {
  static propTypes = {
    /** @private */
    className: PropTypes.string,
    /** Data list to display */
    data: PropTypes.arrayOf(PropTypes.object).isRequired,
    /** Unique identifier for a row in table. */
    rowKey: PropTypes.func,
    /** Auto detect column width. `fixed` layout is faster. */
    layout: PropTypes.oneOf(['auto', 'fixed']),
    /** Sort by columns */
    sort: PropTypes.arrayOf(
      PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.shape({ column: PropTypes.string, order: PropTypes.oneOf(['ASC', 'DESC']) })
      ])
    ),
    /** List of `<Table.Column>` components which declares render behaviour */
    children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]),
    renderRow: PropTypes.func,
    /** Render div with display table. */
    useDiv: PropTypes.bool,
    /** @private */
    _validateChildren({ children }) {
      React.Children.forEach(children, child => {
        if (child.type !== Table.Column) {
          throw new Error('Only Table.Column component can be used as child of Table.')
        }
      })
    }
  }

  static defaultProps = {
    /**
     * @func
     * @param {object} row
     * @param {number} index
     */
    rowKey: (row, index) => index,
    layout: 'fixed',
    renderRow: ({ children }) => React.Children.only(children)
  }

  prepareHead = memoize((children, order) => prepareHead(children, order))
  prepareBody = memoize((head, data, rowKey) => prepareBody(head, data, rowKey))
  prepareSort = memoize(sort => sort.map(column => (typeof column === 'string' ? { column, order: 'ASC' } : column)))

  get head() {
    return this.prepareHead(this.props.children, [])
  }

  get sort() {
    return this.prepareSort(this.props.sort)
  }

  compare = (a, b) => (a < b ? -1 : a > b ? 1 : 0)

  render() {
    const render = data => (
      <TableSimple
        data={this.props.data}
        layout={this.props.layout}
        head={this.head}
        body={this.prepareBody(this.head, data, this.props.rowKey)}
        renderRow={this.props.renderRow}
        useDiv={this.props.useDiv}
      >
        {this.props.children}
      </TableSimple>
    )

    return Array.isArray(this.props.sort) && this.props.sort.length ? (
      <Sortable
        data={this.props.data}
        compareFn={(a, b) => {
          for (const { column, order } of this.sort) {
            const val = this.compare(a[column], b[column])

            if (val !== 0) return order === 'DESC' ? -val : val
          }
          return 0
        }}
      >
        {render}
      </Sortable>
    ) : (
      render(this.props.data)
    )
  }
}

// Sub components.
Table.Column = TableColumn
// Table.Row = TableRow
// Table.Cell = TableCell
// Table.SimpleTable = TableSimple

export default Table
