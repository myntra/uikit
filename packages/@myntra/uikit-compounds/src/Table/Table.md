### Use `accessor`, `key` or `children` to render a cell

``` jsx render editor
<Table data={[
  { id: 1, name: 'Jane Doe', age: 24, status: 'single', visits: 80 },
  { id: 2, name: 'John Doe', age: 26, status: 'single', visits: 120 }
]}>
  <Table.Column label="ID" key="id" />
  <Table.Column label="Name" key="anything" accessor="name" />
  <Table.Column label="Info" key="info">
    <Table.Column label="Age" key="age">
      {({ data }) => <span>{ data.age } years</span>}
    </Table.Column>
    <Table.Column label="Status" key="status" />
  </Table.Column>
  <Table.Column label="Statistics" key="statistics">
    <Table.Column label="Visits" key="visits" accessor={({ visits }) => visits + ' of 200'} />
  </Table.Column>
</Table>
```

### Table using `div`

``` jsx render editor
<Table data={[
  { id: 1, name: 'Jane Doe', age: 24, status: 'single', visits: 80 },
  { id: 2, name: 'John Doe', age: 26, status: 'single', visits: 120 }
]} useDiv>
  <Table.Column label="ID" key="id" />
  <Table.Column label="Name" key="anything" accessor="name" column={{ style: { background: 'rgba(255, 0, 0, .54)' } }} />
  <Table.Column label="Age" key="age">
    {({ data }) => <span>{ data.age } years</span>}
  </Table.Column>
  <Table.Column label="Age (in days)" key="ageInDays">
    {({ data }) => <span>{ data.age } years</span>}
  </Table.Column>
  <Table.Column label="Age (in hours)" key="ageInHours">
    {({ data }) => <span>{ data.age * 365 * 24 } hours</span>}
  </Table.Column>
  <Table.Column label="Age (in minutes)" key="ageInMinutes">
    {({ data }) => <span>{ data.age * 365 * 24 * 60 } minutes</span>}
  </Table.Column>
  <Table.Column label="Age (in seconds)" key="ageInSeconds">
    {({ data }) => <span>{ data.age * 365 * 24 * 60 * 60 } seconds</span>}
  </Table.Column>
  <Table.Column label="Age (in milliseconds)" key="ageInMilliseconds">
    {({ data }) => <span>{ data.age * 365 * 24 * 60 * 60 * 1000 } milliseconds</span>}
  </Table.Column>
  <Table.Column label="Status" key="status" />
  <Table.Column label="Visits" key="visits" accessor={({ visits }) => visits + ' of 200'} />
  <Table.Column label="Visits (%)" key="visitsPercentage" accessor={({ visits }) => visits / 2  + '%'} />
  <Table.Column label="New Visits" key="visitsNew" accessor={({ visits }) => visits / 2} />
</Table>
```

### Collapsible Rows

``` jsx render editor
<Accordion active={null}>
  <Table
    data={[
      { id: 1, name: 'Jane Doe' },
      { id: 2, name: 'John Doe' },
    ]}
    useDiv
    renderRow={({ children, index, data, head, WithSentinel }) => (
      head ? children :
      <Accordion.Item title={children} renderWrapper={({ children }) => [children]}>
        <WithSentinel>
          <p style={{ padding: '16px' }}>
            I am {data && data.name} at position {index}.
          </p>
        </WithSentinel>
      </Accordion.Item>
    )}
  >
    <Table.Column label="ID" key="id" />
    <Table.Column label="Name" key="name" />
  </Table>
</Accordion>
```
