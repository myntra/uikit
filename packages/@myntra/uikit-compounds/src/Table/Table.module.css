.wrapper {
  position: relative;
}

.scroller {
  overflow-x: auto;
  overflow-y: visible;
}

.wrapper .freeze {
  position: absolute;
}

.table {
  min-width: 100%;
  margin-bottom: 1rem;
  background-color: transparent;
  border-spacing: 0;
  border-collapse: collapse;
  border: 1px solid #dee2e6;
}

div.table {
  display: table;
}

.table.fixed {
  table-layout: fixed;
}

.table .thead {
  border-bottom: 2px solid #dee2e6;
}

.table div.thead {
  width: 100%;
  display: table-header-group;
}

.table div.tbody {
  width: 100%;
  display: table-row-group;
}

.table .row {
  display: table-row;
  border-bottom: 1px solid #dee2e6;
}

.table .col {
  display: table-cell;
  border: 1px solid #dee2e6;
  padding: 0.75rem;
  white-space: nowrap;
  vertical-align: middle;
}

.table .col.head {
  background-color: #f3f3f3;
  vertical-align: bottom;
}

