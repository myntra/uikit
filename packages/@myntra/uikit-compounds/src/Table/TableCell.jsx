import React from 'react'
import PropTypes from 'prop-types'
import { classnames } from '@myntra/uikit-utils'
import styles from './Table.module.css'

/**
 Render a table cell

 @since 0.3.0
 @status EXPERIMENTAL
 @example
<table>
  <tbody>
    <tr>
      <Table.Cell head colSpan={2}>
        Foo
      </Table.Cell>
    </tr>
  </tbody>
</table>
 */
function Cell({ head, children, freeze, className, useDiv, ...props }) {
  const Td = useDiv ? 'div' : head ? 'th' : 'td'

  return (
    <Td {...props} className={classnames('col', className, { freeze, head }).use(styles)}>
      {children}
    </Td>
  )
}

Cell.propTypes = {
  /** @private */
  className: PropTypes.string,
  /** Header cell (`<th>`) */
  head: PropTypes.bool,
  /** Freeze cell */
  freeze: PropTypes.bool,
  /** Number of table columns to use */
  colSpan: PropTypes.number,
  /** Contents of the cell */
  children: PropTypes.any,
  /** Render div with display table-cell. */
  useDiv: PropTypes.bool
}

export default Cell
