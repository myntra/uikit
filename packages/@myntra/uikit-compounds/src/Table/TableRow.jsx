import React from 'react'
import PropTypes from 'prop-types'
import styles from './Table.module.css'
import { classnames } from '@myntra/uikit-utils'

/**
 Render a table row

 @since 0.3.0
 @status EXPERIMENTAL
 @example
<Table.Row>
  <td>Foo</td>
  <td>Bar</td>
</Table.Row>
 */
function Row({ children, useDiv, className, ...props }) {
  const Tr = useDiv ? 'div' : 'tr'

  return (
    <Tr className={classnames('row', className).use(styles)} {...props}>
      {children}
    </Tr>
  )
}

Row.propTypes = {
  /** @private */
  className: PropTypes.string,
  /** Contents of the cell */
  children: PropTypes.any,
  /** Render div with display table-row. */
  useDiv: PropTypes.bool
}

export default Row
