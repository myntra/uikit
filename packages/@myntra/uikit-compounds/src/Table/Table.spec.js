import React from 'react'
import { mount } from 'enzyme'
import Table from './Table'

it('should render correctly', () => {
  const wrapper = mount(
    <Table data={[{ id: 1, firstName: 'John', lastName: 'Doe', content: 'foo' }]}>
      <Table.Column key="id" label="ID" />
      <Table.Column key="name" label="Name" accessor={({ firstName, lastName }) => `${firstName} ${lastName}`} />
      <Table.Column key="content" label="Content">
        {({ data }) => <span>{data.content}</span>}
      </Table.Column>
    </Table>
  )

  expect(wrapper).toMatchSnapshot()
})

it('should render sorted rows', () => {
  const wrapper = mount(
    <Table
      data={[
        { id: 1, firstName: 'John', lastName: 'Doe', content: 'foo' },
        { id: 2, firstName: 'Jane', lastName: 'Doe', content: 'foo' }
      ]}
      sort={['content']}
    >
      <Table.Column key="id" label="ID" />
      <Table.Column key="name" label="Name" accessor={({ firstName, lastName }) => `${firstName} ${lastName}`} />
      <Table.Column key="content" label="Content">
        {({ data }) => <span>{data.content}</span>}
      </Table.Column>
    </Table>
  )

  const rows = wrapper.find('tbody').find('tr')

  expect(wrapper).toMatchSnapshot()

  expect(rows).toHaveLength(2)

  /* eslint-disable prettier/prettier */
  expect(rows.at(0).find('td').first().text()).toEqual('1')
  expect(rows.at(1).find('td').first().text()).toEqual('2')
  /* eslint-enable prettier/prettier */
})

it('should render reverse sorted rows', () => {
  const wrapper = mount(
    <Table
      data={[
        { id: 1, firstName: 'John', lastName: 'Doe', content: 'foo' },
        { id: 2, firstName: 'Jane', lastName: 'Doe', content: 'bar' },
        { id: 3, firstName: 'Jane', lastName: 'Do', content: 'baz' }
      ]}
      sort={[{ column: 'firstName', order: 'ASC' }, { column: 'content', order: 'DESC' }]}
    >
      <Table.Column key="id" label="ID" />
      <Table.Column key="name" label="Name" accessor={({ firstName, lastName }) => `${firstName} ${lastName}`} />
      <Table.Column key="content" label="Content">
        {({ data }) => <span>{data.content}</span>}
      </Table.Column>
    </Table>
  )

  const rows = wrapper.find('tbody').find('tr')

  expect(wrapper).toMatchSnapshot()

  expect(rows).toHaveLength(3)

  /* eslint-disable prettier/prettier */
  expect(rows.at(0).find('td').first().text()).toEqual('3')
  expect(rows.at(1).find('td').first().text()).toEqual('2')
  expect(rows.at(2).find('td').first().text()).toEqual('1')
  /* eslint-enable prettier/prettier */
})

it('should throw error on unknown children', () => {
  const spy = jest.spyOn(console, 'error').mockImplementation(() => {})
  mount(
    <Table data={[]}>
      <Table.Column key="id" label="ID" />
      <div key="1" />
    </Table>
  )

  expect(spy).toHaveBeenCalled()
})
