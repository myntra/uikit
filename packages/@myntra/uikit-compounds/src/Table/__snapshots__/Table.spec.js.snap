// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<Table
  data={
    Array [
      Object {
        "content": "foo",
        "firstName": "John",
        "id": 1,
        "lastName": "Doe",
      },
    ]
  }
  layout="fixed"
  renderRow={[Function]}
  rowKey={[Function]}
>
  <TableSimple
    body={
      Object {
        "rows": Array [
          Object {
            "columns": Array [
              Object {
                "children": 1,
                "key": "id",
              },
              Object {
                "children": "John Doe",
                "key": "name",
              },
              Object {
                "children": <span>
                  foo
                </span>,
                "key": "content",
              },
            ],
            "id": 0,
          },
        ],
      }
    }
    data={
      Array [
        Object {
          "content": "foo",
          "firstName": "John",
          "id": 1,
          "lastName": "Doe",
        },
      ]
    }
    head={
      Object {
        "colgroup": Array [
          Object {
            "children": "ID",
            "colSpan": 1,
            "key": "id",
            "rowSpan": 1,
          },
          Object {
            "children": "Name",
            "colSpan": 1,
            "key": "name",
            "rowSpan": 1,
          },
          Object {
            "children": "Content",
            "colSpan": 1,
            "key": "content",
            "rowSpan": 1,
          },
        ],
        "columns": Object {
          "content": Object {
            "children": Array [],
            "depth": 1,
            "key": "content",
            "props": Object {
              "children": "Content",
              "colSpan": 1,
              "key": "content",
              "rowSpan": 1,
            },
            "render": [Function],
          },
          "id": Object {
            "children": Array [],
            "depth": 1,
            "key": "id",
            "props": Object {
              "children": "ID",
              "colSpan": 1,
              "key": "id",
              "rowSpan": 1,
            },
            "render": [Function],
          },
          "name": Object {
            "children": Array [],
            "depth": 1,
            "key": "name",
            "props": Object {
              "children": "Name",
              "colSpan": 1,
              "key": "name",
              "rowSpan": 1,
            },
            "render": [Function],
          },
        },
        "order": Array [
          Object {
            "children": Array [],
            "depth": 1,
            "key": "id",
            "props": Object {
              "children": "ID",
              "colSpan": 1,
              "key": "id",
              "rowSpan": 1,
            },
            "render": [Function],
          },
          Object {
            "children": Array [],
            "depth": 1,
            "key": "name",
            "props": Object {
              "children": "Name",
              "colSpan": 1,
              "key": "name",
              "rowSpan": 1,
            },
            "render": [Function],
          },
          Object {
            "children": Array [],
            "depth": 1,
            "key": "content",
            "props": Object {
              "children": "Content",
              "colSpan": 1,
              "key": "content",
              "rowSpan": 1,
            },
            "render": [Function],
          },
        ],
        "rows": Array [
          Object {
            "columns": Array [
              Object {
                "children": "ID",
                "colSpan": 1,
                "key": "id",
                "rowSpan": 1,
              },
              Object {
                "children": "Name",
                "colSpan": 1,
                "key": "name",
                "rowSpan": 1,
              },
              Object {
                "children": "Content",
                "colSpan": 1,
                "key": "content",
                "rowSpan": 1,
              },
            ],
            "id": 0,
          },
        ],
      }
    }
    layout="fixed"
    renderRow={[Function]}
  >
    <div
      className="wrapper"
    >
      <div
        className="scroller"
      >
        <table
          className="table fixed"
        >
          <thead
            className="thead"
          >
            <renderRow
              key="0"
            >
              <Row>
                <tr
                  className="row"
                >
                  <Cell
                    colSpan={1}
                    head={true}
                    key="id"
                    rowSpan={1}
                  >
                    <th
                      className="col head"
                      colSpan={1}
                      rowSpan={1}
                    >
                      ID
                    </th>
                  </Cell>
                  <Cell
                    colSpan={1}
                    head={true}
                    key="name"
                    rowSpan={1}
                  >
                    <th
                      className="col head"
                      colSpan={1}
                      rowSpan={1}
                    >
                      Name
                    </th>
                  </Cell>
                  <Cell
                    colSpan={1}
                    head={true}
                    key="content"
                    rowSpan={1}
                  >
                    <th
                      className="col head"
                      colSpan={1}
                      rowSpan={1}
                    >
                      Content
                    </th>
                  </Cell>
                </tr>
              </Row>
            </renderRow>
          </thead>
          <tbody
            className="tbody"
          >
            <renderRow
              key="0"
            >
              <Row>
                <tr
                  className="row"
                >
                  <Cell
                    key="id"
                  >
                    <td
                      className="col"
                    >
                      1
                    </td>
                  </Cell>
                  <Cell
                    key="name"
                  >
                    <td
                      className="col"
                    >
                      John Doe
                    </td>
                  </Cell>
                  <Cell
                    key="content"
                  >
                    <td
                      className="col"
                    >
                      <span>
                        foo
                      </span>
                    </td>
                  </Cell>
                </tr>
              </Row>
            </renderRow>
          </tbody>
        </table>
      </div>
    </div>
  </TableSimple>
</Table>
`;

exports[`should render reverse sorted rows 1`] = `
<Table
  data={
    Array [
      Object {
        "content": "foo",
        "firstName": "John",
        "id": 1,
        "lastName": "Doe",
      },
      Object {
        "content": "bar",
        "firstName": "Jane",
        "id": 2,
        "lastName": "Doe",
      },
      Object {
        "content": "baz",
        "firstName": "Jane",
        "id": 3,
        "lastName": "Do",
      },
    ]
  }
  layout="fixed"
  renderRow={[Function]}
  rowKey={[Function]}
  sort={
    Array [
      Object {
        "column": "firstName",
        "order": "ASC",
      },
      Object {
        "column": "content",
        "order": "DESC",
      },
    ]
  }
>
  <Sortable
    compareFn={[Function]}
    data={
      Array [
        Object {
          "content": "foo",
          "firstName": "John",
          "id": 1,
          "lastName": "Doe",
        },
        Object {
          "content": "bar",
          "firstName": "Jane",
          "id": 2,
          "lastName": "Doe",
        },
        Object {
          "content": "baz",
          "firstName": "Jane",
          "id": 3,
          "lastName": "Do",
        },
      ]
    }
  >
    <TableSimple
      body={
        Object {
          "rows": Array [
            Object {
              "columns": Array [
                Object {
                  "children": 3,
                  "key": "id",
                },
                Object {
                  "children": "Jane Do",
                  "key": "name",
                },
                Object {
                  "children": <span>
                    baz
                  </span>,
                  "key": "content",
                },
              ],
              "id": 0,
            },
            Object {
              "columns": Array [
                Object {
                  "children": 2,
                  "key": "id",
                },
                Object {
                  "children": "Jane Doe",
                  "key": "name",
                },
                Object {
                  "children": <span>
                    bar
                  </span>,
                  "key": "content",
                },
              ],
              "id": 1,
            },
            Object {
              "columns": Array [
                Object {
                  "children": 1,
                  "key": "id",
                },
                Object {
                  "children": "John Doe",
                  "key": "name",
                },
                Object {
                  "children": <span>
                    foo
                  </span>,
                  "key": "content",
                },
              ],
              "id": 2,
            },
          ],
        }
      }
      data={
        Array [
          Object {
            "content": "foo",
            "firstName": "John",
            "id": 1,
            "lastName": "Doe",
          },
          Object {
            "content": "bar",
            "firstName": "Jane",
            "id": 2,
            "lastName": "Doe",
          },
          Object {
            "content": "baz",
            "firstName": "Jane",
            "id": 3,
            "lastName": "Do",
          },
        ]
      }
      head={
        Object {
          "colgroup": Array [
            Object {
              "children": "ID",
              "colSpan": 1,
              "key": "id",
              "rowSpan": 1,
            },
            Object {
              "children": "Name",
              "colSpan": 1,
              "key": "name",
              "rowSpan": 1,
            },
            Object {
              "children": "Content",
              "colSpan": 1,
              "key": "content",
              "rowSpan": 1,
            },
          ],
          "columns": Object {
            "content": Object {
              "children": Array [],
              "depth": 1,
              "key": "content",
              "props": Object {
                "children": "Content",
                "colSpan": 1,
                "key": "content",
                "rowSpan": 1,
              },
              "render": [Function],
            },
            "id": Object {
              "children": Array [],
              "depth": 1,
              "key": "id",
              "props": Object {
                "children": "ID",
                "colSpan": 1,
                "key": "id",
                "rowSpan": 1,
              },
              "render": [Function],
            },
            "name": Object {
              "children": Array [],
              "depth": 1,
              "key": "name",
              "props": Object {
                "children": "Name",
                "colSpan": 1,
                "key": "name",
                "rowSpan": 1,
              },
              "render": [Function],
            },
          },
          "order": Array [
            Object {
              "children": Array [],
              "depth": 1,
              "key": "id",
              "props": Object {
                "children": "ID",
                "colSpan": 1,
                "key": "id",
                "rowSpan": 1,
              },
              "render": [Function],
            },
            Object {
              "children": Array [],
              "depth": 1,
              "key": "name",
              "props": Object {
                "children": "Name",
                "colSpan": 1,
                "key": "name",
                "rowSpan": 1,
              },
              "render": [Function],
            },
            Object {
              "children": Array [],
              "depth": 1,
              "key": "content",
              "props": Object {
                "children": "Content",
                "colSpan": 1,
                "key": "content",
                "rowSpan": 1,
              },
              "render": [Function],
            },
          ],
          "rows": Array [
            Object {
              "columns": Array [
                Object {
                  "children": "ID",
                  "colSpan": 1,
                  "key": "id",
                  "rowSpan": 1,
                },
                Object {
                  "children": "Name",
                  "colSpan": 1,
                  "key": "name",
                  "rowSpan": 1,
                },
                Object {
                  "children": "Content",
                  "colSpan": 1,
                  "key": "content",
                  "rowSpan": 1,
                },
              ],
              "id": 0,
            },
          ],
        }
      }
      layout="fixed"
      renderRow={[Function]}
    >
      <div
        className="wrapper"
      >
        <div
          className="scroller"
        >
          <table
            className="table fixed"
          >
            <thead
              className="thead"
            >
              <renderRow
                key="0"
              >
                <Row>
                  <tr
                    className="row"
                  >
                    <Cell
                      colSpan={1}
                      head={true}
                      key="id"
                      rowSpan={1}
                    >
                      <th
                        className="col head"
                        colSpan={1}
                        rowSpan={1}
                      >
                        ID
                      </th>
                    </Cell>
                    <Cell
                      colSpan={1}
                      head={true}
                      key="name"
                      rowSpan={1}
                    >
                      <th
                        className="col head"
                        colSpan={1}
                        rowSpan={1}
                      >
                        Name
                      </th>
                    </Cell>
                    <Cell
                      colSpan={1}
                      head={true}
                      key="content"
                      rowSpan={1}
                    >
                      <th
                        className="col head"
                        colSpan={1}
                        rowSpan={1}
                      >
                        Content
                      </th>
                    </Cell>
                  </tr>
                </Row>
              </renderRow>
            </thead>
            <tbody
              className="tbody"
            >
              <renderRow
                key="0"
              >
                <Row>
                  <tr
                    className="row"
                  >
                    <Cell
                      key="id"
                    >
                      <td
                        className="col"
                      >
                        3
                      </td>
                    </Cell>
                    <Cell
                      key="name"
                    >
                      <td
                        className="col"
                      >
                        Jane Do
                      </td>
                    </Cell>
                    <Cell
                      key="content"
                    >
                      <td
                        className="col"
                      >
                        <span>
                          baz
                        </span>
                      </td>
                    </Cell>
                  </tr>
                </Row>
              </renderRow>
              <renderRow
                key="1"
              >
                <Row>
                  <tr
                    className="row"
                  >
                    <Cell
                      key="id"
                    >
                      <td
                        className="col"
                      >
                        2
                      </td>
                    </Cell>
                    <Cell
                      key="name"
                    >
                      <td
                        className="col"
                      >
                        Jane Doe
                      </td>
                    </Cell>
                    <Cell
                      key="content"
                    >
                      <td
                        className="col"
                      >
                        <span>
                          bar
                        </span>
                      </td>
                    </Cell>
                  </tr>
                </Row>
              </renderRow>
              <renderRow
                key="2"
              >
                <Row>
                  <tr
                    className="row"
                  >
                    <Cell
                      key="id"
                    >
                      <td
                        className="col"
                      >
                        1
                      </td>
                    </Cell>
                    <Cell
                      key="name"
                    >
                      <td
                        className="col"
                      >
                        John Doe
                      </td>
                    </Cell>
                    <Cell
                      key="content"
                    >
                      <td
                        className="col"
                      >
                        <span>
                          foo
                        </span>
                      </td>
                    </Cell>
                  </tr>
                </Row>
              </renderRow>
            </tbody>
          </table>
        </div>
      </div>
    </TableSimple>
  </Sortable>
</Table>
`;

exports[`should render sorted rows 1`] = `
<Table
  data={
    Array [
      Object {
        "content": "foo",
        "firstName": "John",
        "id": 1,
        "lastName": "Doe",
      },
      Object {
        "content": "foo",
        "firstName": "Jane",
        "id": 2,
        "lastName": "Doe",
      },
    ]
  }
  layout="fixed"
  renderRow={[Function]}
  rowKey={[Function]}
  sort={
    Array [
      "content",
    ]
  }
>
  <Sortable
    compareFn={[Function]}
    data={
      Array [
        Object {
          "content": "foo",
          "firstName": "John",
          "id": 1,
          "lastName": "Doe",
        },
        Object {
          "content": "foo",
          "firstName": "Jane",
          "id": 2,
          "lastName": "Doe",
        },
      ]
    }
  >
    <TableSimple
      body={
        Object {
          "rows": Array [
            Object {
              "columns": Array [
                Object {
                  "children": 1,
                  "key": "id",
                },
                Object {
                  "children": "John Doe",
                  "key": "name",
                },
                Object {
                  "children": <span>
                    foo
                  </span>,
                  "key": "content",
                },
              ],
              "id": 0,
            },
            Object {
              "columns": Array [
                Object {
                  "children": 2,
                  "key": "id",
                },
                Object {
                  "children": "Jane Doe",
                  "key": "name",
                },
                Object {
                  "children": <span>
                    foo
                  </span>,
                  "key": "content",
                },
              ],
              "id": 1,
            },
          ],
        }
      }
      data={
        Array [
          Object {
            "content": "foo",
            "firstName": "John",
            "id": 1,
            "lastName": "Doe",
          },
          Object {
            "content": "foo",
            "firstName": "Jane",
            "id": 2,
            "lastName": "Doe",
          },
        ]
      }
      head={
        Object {
          "colgroup": Array [
            Object {
              "children": "ID",
              "colSpan": 1,
              "key": "id",
              "rowSpan": 1,
            },
            Object {
              "children": "Name",
              "colSpan": 1,
              "key": "name",
              "rowSpan": 1,
            },
            Object {
              "children": "Content",
              "colSpan": 1,
              "key": "content",
              "rowSpan": 1,
            },
          ],
          "columns": Object {
            "content": Object {
              "children": Array [],
              "depth": 1,
              "key": "content",
              "props": Object {
                "children": "Content",
                "colSpan": 1,
                "key": "content",
                "rowSpan": 1,
              },
              "render": [Function],
            },
            "id": Object {
              "children": Array [],
              "depth": 1,
              "key": "id",
              "props": Object {
                "children": "ID",
                "colSpan": 1,
                "key": "id",
                "rowSpan": 1,
              },
              "render": [Function],
            },
            "name": Object {
              "children": Array [],
              "depth": 1,
              "key": "name",
              "props": Object {
                "children": "Name",
                "colSpan": 1,
                "key": "name",
                "rowSpan": 1,
              },
              "render": [Function],
            },
          },
          "order": Array [
            Object {
              "children": Array [],
              "depth": 1,
              "key": "id",
              "props": Object {
                "children": "ID",
                "colSpan": 1,
                "key": "id",
                "rowSpan": 1,
              },
              "render": [Function],
            },
            Object {
              "children": Array [],
              "depth": 1,
              "key": "name",
              "props": Object {
                "children": "Name",
                "colSpan": 1,
                "key": "name",
                "rowSpan": 1,
              },
              "render": [Function],
            },
            Object {
              "children": Array [],
              "depth": 1,
              "key": "content",
              "props": Object {
                "children": "Content",
                "colSpan": 1,
                "key": "content",
                "rowSpan": 1,
              },
              "render": [Function],
            },
          ],
          "rows": Array [
            Object {
              "columns": Array [
                Object {
                  "children": "ID",
                  "colSpan": 1,
                  "key": "id",
                  "rowSpan": 1,
                },
                Object {
                  "children": "Name",
                  "colSpan": 1,
                  "key": "name",
                  "rowSpan": 1,
                },
                Object {
                  "children": "Content",
                  "colSpan": 1,
                  "key": "content",
                  "rowSpan": 1,
                },
              ],
              "id": 0,
            },
          ],
        }
      }
      layout="fixed"
      renderRow={[Function]}
    >
      <div
        className="wrapper"
      >
        <div
          className="scroller"
        >
          <table
            className="table fixed"
          >
            <thead
              className="thead"
            >
              <renderRow
                key="0"
              >
                <Row>
                  <tr
                    className="row"
                  >
                    <Cell
                      colSpan={1}
                      head={true}
                      key="id"
                      rowSpan={1}
                    >
                      <th
                        className="col head"
                        colSpan={1}
                        rowSpan={1}
                      >
                        ID
                      </th>
                    </Cell>
                    <Cell
                      colSpan={1}
                      head={true}
                      key="name"
                      rowSpan={1}
                    >
                      <th
                        className="col head"
                        colSpan={1}
                        rowSpan={1}
                      >
                        Name
                      </th>
                    </Cell>
                    <Cell
                      colSpan={1}
                      head={true}
                      key="content"
                      rowSpan={1}
                    >
                      <th
                        className="col head"
                        colSpan={1}
                        rowSpan={1}
                      >
                        Content
                      </th>
                    </Cell>
                  </tr>
                </Row>
              </renderRow>
            </thead>
            <tbody
              className="tbody"
            >
              <renderRow
                key="0"
              >
                <Row>
                  <tr
                    className="row"
                  >
                    <Cell
                      key="id"
                    >
                      <td
                        className="col"
                      >
                        1
                      </td>
                    </Cell>
                    <Cell
                      key="name"
                    >
                      <td
                        className="col"
                      >
                        John Doe
                      </td>
                    </Cell>
                    <Cell
                      key="content"
                    >
                      <td
                        className="col"
                      >
                        <span>
                          foo
                        </span>
                      </td>
                    </Cell>
                  </tr>
                </Row>
              </renderRow>
              <renderRow
                key="1"
              >
                <Row>
                  <tr
                    className="row"
                  >
                    <Cell
                      key="id"
                    >
                      <td
                        className="col"
                      >
                        2
                      </td>
                    </Cell>
                    <Cell
                      key="name"
                    >
                      <td
                        className="col"
                      >
                        Jane Doe
                      </td>
                    </Cell>
                    <Cell
                      key="content"
                    >
                      <td
                        className="col"
                      >
                        <span>
                          foo
                        </span>
                      </td>
                    </Cell>
                  </tr>
                </Row>
              </renderRow>
            </tbody>
          </table>
        </div>
      </div>
    </TableSimple>
  </Sortable>
</Table>
`;
